### ISSUE TITLE
Pyright reports that class is missing __aenter__ when it isn’t 

### ISSUE THREAD

**Issue by adriangb** [microsoft/pyright#6849]
```python
from psycopg_pool import AsyncConnectionPool

async def main() -> None:
    async with AsyncConnectionPool() as pool:
        pass
```

```
test.py
  test.py:5:16 - error: Object of type "AsyncConnectionPool[AsyncConnection[TupleRow]]" cannot be used with "with" because it does not implement __aenter__ (reportGeneralTypeIssues)
  test.py:5:41 - error: Type of "pool" is unknown (reportUnknownVariableType)
  test.py:5:41 - error: Variable "pool" is not accessed (reportUnusedVariable)
3 errors, 0 warnings, 0 informations 
```

pyright==1.1.343
psycopg==3.1.16
psycopg-binary==3.1.16
psycopg-pool==3.2.0

I [reported the issue on psycopg](https://github.com/psycopg/psycopg/issues/708) but it seems like pyright may be at fault, or at least the error message is unclear if not incorrect.

**Comment by erictraut** [comment]
Pyright is correct to generate an error here, but I agree the error could be clearer. The problem is that while there is an `__aenter__` present, its `self` parameter is annotated such that it cannot be bound. This stems from the fact that it uses a TypeVar that is defined in an illegal way (using an upper bound that is generic). The library maintainers will need to make changes if they want this library to work as intended in type checkers that follow the typing standard.

I've improved the error message in pyright in this case so the underlying cause of the error is easier to diagnose.

**Comment by adriangb** [comment]
Excellent thank you Eric. Feel free to close this issue.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.344, which I just published. It will also be included in a future release of pylance.
